openvdb_ax/openvdb_ax/ast/Tokens.h

```diff
@@ -0,0 +1,364 @@
+template<> inline CoreType tokenFromType<math::Vec3<double>>()   { return VEC3D; }
+template<> inline CoreType tokenFromType<math::Vec4<int32_t>>()  { return VEC4I; }
+template<> inline CoreType tokenFromType<math::Vec4<float>>()    { return VEC4F; }
+template<> inline CoreType tokenFromType<math::Vec4<double>>()   { return VEC4D; }
+template<> inline CoreType tokenFromType<math::Mat3<float>>()    { return MAT3F; }
+template<> inline CoreType tokenFromType<math::Mat3<double>>()   { return MAT3D; }
+template<> inline CoreType tokenFromType<math::Mat4<float>>()    { return MAT4F; }
+template<> inline CoreType tokenFromType<math::Mat4<double>>()   { return MAT4D; }
+template<> inline CoreType tokenFromType<std::string>()          { return STRING; }
+
+enum OperatorToken
+{
+    ////////////////////////////////////////////////////////////////
+    ///  ARITHMETIC
+    ////////////////////////////////////////////////////////////////
+    PLUS = 0,
+    MINUS,
+    MULTIPLY,
+    DIVIDE,
+    MODULO,
+    ////////////////////////////////////////////////////////////////
+    ///  LOGICAL
+    ////////////////////////////////////////////////////////////////
+    AND,
+    OR,
+    NOT,
+    ////////////////////////////////////////////////////////////////
+    ///  RELATIONAL
+    ////////////////////////////////////////////////////////////////
+    EQUALSEQUALS,
+    NOTEQUALS,
+    MORETHAN,
+    LESSTHAN,
+    MORETHANOREQUAL,
+    LESSTHANOREQUAL,
+    ////////////////////////////////////////////////////////////////
+    ///  BITWISE
+    ////////////////////////////////////////////////////////////////
+    SHIFTLEFT,
+    SHIFTRIGHT,
+    BITAND,
+    BITOR,
+    BITXOR,
+    BITNOT,
+    ////////////////////////////////////////////////////////////////
+    ///  ASSIGNMENT
+    ////////////////////////////////////////////////////////////////
+    EQUALS,
+    PLUSEQUALS,
+    MINUSEQUALS,
+    MULTIPLYEQUALS,
+    DIVIDEEQUALS,
+    MODULOEQUALS,
+    SHIFTLEFTEQUALS,
+    SHIFTRIGHTEQUALS,
+    BITANDEQUALS,
+    BITXOREQUALS,
+    BITOREQUALS
+};
+
+enum OperatorType
+{
+    ARITHMETIC = 0,
+    LOGICAL,
+    RELATIONAL,
+    BITWISE,
+    ASSIGNMENT,
+    UNKNOWN_OPERATOR
+};
+
+inline OperatorType operatorType(const OperatorToken token)
+{
+    const size_t idx = static_cast<size_t>(token);
+    if (idx <= static_cast<size_t>(MODULO))          return ARITHMETIC;
+    if (idx <= static_cast<size_t>(NOT))             return LOGICAL;
+    if (idx <= static_cast<size_t>(LESSTHANOREQUAL)) return RELATIONAL;
+    if (idx <= static_cast<size_t>(BITNOT))          return BITWISE;
+    if (idx <= static_cast<size_t>(BITOREQUALS))     return ASSIGNMENT;
+    return UNKNOWN_OPERATOR;
+}
+
+inline OperatorToken operatorTokenFromName(const std::string& name)
+{
+    if (name == "+")    return PLUS;
+    if (name == "-")    return MINUS;
+    if (name == "*")    return MULTIPLY;
+    if (name == "/")    return DIVIDE;
+    if (name == "%")    return MODULO;
```

Comment:
I've gone down a bit of a rabbit hole with this. It seems that most languages I've looked into (bar c) implement a "floored modulo" operation, where the divisor in `a % b` is assumed to be positive - essentially `((a % b) + b) % b;` (assuming % has been implemented as a "truncated modulo"). As I understand it, this still isn't a true euclidean mod, which should always produce positive results. i.e.
```
x = a%b;
if (x < 0) {
   if (b > 0) x = x + b;
   else         x = x - b;
}
```
This is essentially the most performant implementation (using truncated mod) of a pure euclidean mode where as the floored impl can be improved to:

```
   (a < 0 ? ((a % b) + b) % b : a % b);
```

This version preserves the sign of the divisor, however it's debatable how important or how "correct" propagating this information is. In terms of its implementation and regularity, f-mod seems to be the best balance between the three, although e-mod covers all cases in the most uniform way. As well as speed and ... "precision?" we also care about compatibility. The fact that AX/VEX is so similar to C concerns me a bit, though within applications like Houdini users are perhaps expected to be more familiar with python. This is something I've never considered before... Part of me is now wondering if it should even be supported in favour of explicit functions :cold_sweat: 

http://norayr.am/papers/divmodnote.pdf
https://core.ac.uk/download/pdf/187613369.pdf

Reply:
One last concern with this is that python, which implements floored mod, also implements floored division. This means that in both C and python, the following expressions are true:

```
  (-7/3)*3 + (-7%3) == -7
  (-7/-3)*-3 + (-7%-3) == -7
  (7/-3)*-3 + (7%-3) == 7
```

These are **not** true in VEX. Are we sure this is the right decision? I would think that keeping these properties and relationship is more important than the behaviour of mod. @jmlait?

Reply:
That is a good point about the non-floored division not matching the floored mod.  Ideally, those would match.

But I disagree that maintaining that invariant is the most important thing.  The most important thing is for people writing selection functions with mod that they get an expected range.

int choice = something % nchoice;

is a very common idiom that fails utterly when something is signed.  I've seen people forget about sign in that case too many times to be complacent about this.

The other use of mod is for wrapping/tiling.  In this case, the lack of floored division means that to get the tile number you can't just use integer division.   But I'll note that maintaining that invariant won't help someone trying to convert a coord into a tile/tileoffset:

int tile = x / resx;
int tileoffset = x % resx;

The invariant that this will rebuild to the original x with
x = tile * resx + tileoffset;
is rather moot - you have a negative tile offset in the C'style, and a repeated tile index around 0, so you already have serious bugs with the behaviour at zero.

The break in the division invariant is annoying, but is also not without precedent.  Ada also breaks mod from div, likely for this very reason - the variant of mod that preserves the second-term's sign is the safer choice to avoid bugs.  They interestingly separate the concept of mod from the concept of remainder - remainder should satisfy that property, mod does not have to.

Support % vs explicit functions is a bit moot, you still have to argue what the `mod` function does :>  We do end up with SYSsafemod() (C-style) and SYSwrapmod() (Python style) in Houdini.

Reply:
Ad discussed, I've made this change! AX now performs floored mod

Reply:
As discussed I've made this change! AX now performs floored mod